synapse/push/bulk_push_rule_evaluator.py

```diff
@@ -466,6 +466,27 @@ def _flatten_dict(
     prefix: Optional[List[str]] = None,
     result: Optional[Dict[str, str]] = None,
 ) -> Dict[str, str]:
+    """
+    Given a JSON dictionary (or event) which might contain sub dictionaries,
+    flatten it into a single layer dictionary by combining the keys & sub-keys.
+
```

Comment:
Should note that non-string items are dropped; doesn't sound obvious from the name